keps/sig-node/20190825-exec-as-specified-user.md

```diff
@@ -0,0 +1,129 @@
+replaces:
+  - "n/a"
+superseded-by:
+  - "n/a"
+---
+
+# Exec as specified user
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+- [Design Details](#design-details)
+  - [API Specification](#api-specification)
+  - [Library Specification](#library-specification)
+- [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Summary
+
+As a Kubernetes User, we should be able to specify username for the containers when doing exec, similar to how docker allows that using docker exec options `-u`, 
+```
+-u, --user="" Username or UID (format: <name|uid>[:<group|gid>]) format
+```
+
+## Motivation
+
+The root problem is lacking of standard description for the `exec` operation.
+
+OCI didn't introduce `exec` as a regular operation for container runtime[2],
+
+so when Kubernetes announce [Container Runtime Interface(CRI)](https://kubernetes.io/blog/2016/12/container-runtime-interface-cri-in-kubernetes/) in release 1.5, it became the de-facto standard of `exec`, without a `user` option.
+
+
+>  In the Kubernetes 1.5 release, we are proud to introduce the Container Runtime Interface (CRI) – a plugin interface which enables kubelet to use a wide variety of container runtimes, without the need to recompile. CRI consists of a protocol buffers and gRPC API, and libraries, with additional specifications and tools under active development. CRI is being released as Alpha in Kubernetes 1.5.
+
+But There are many container runtime implementation, e.g Docker, Kata, gVisior, most of which provide a way to enter the container as a special `user`.
+
+It is because the `exec` is basically to start a process inside the container runtime, and the [`process`](https://github.com/opencontainers/runtime-spec/blob/master/specs-go/config.go#L39) has a user definition, which is required.
+
+> User specifies specific user (and group) information for the container process
+
+It brings convenience to debugging, and users are used to it. Adding a `user` option with `exec`  in Kubernetes seems reasonable.
+
+### Goals
+
+1. Provide the ability to specify the username for a exec operation inside a container
+
+## Proposal
+
+### User Stories [optional]
+
+As a Kubernetes User, I should be able to control how people can enter the container, by default kubernetes use the default user in docker image.
+
+#### Story 1
+By providing the `user` option for `exec`, users can use their own ID to enter the container and exercise their respective rights.
+
+```
+-u, --user="" Username or UID (format: <string>]) format
+```
+
+Since the container does not have a uniform `exec` format, the format of `user` should be just a string.
+
+#### Story 2
+By providing the `user` option for `exec`, user can do the auit inside the container.
+
+## Design Details
```

Comment:
There’s a lot of additional security protection which is required to make arbitrary exec safe, especially by username (which is only mappable inside a container).

Anyone using PSP or similar systems would need a way to control which users are allowed to exec.

The use of username is contrary to how kubernetes pods run so it needs substantially more justification.

The design would also have to take into account how windows exec would support exec, and how runtimes that don’t support exec changes would communicate that to users.